### PR TITLE
Add some more type annotations

### PR DESCRIPTION
**Pull request by r-barnes**
Adds additional type annotations

### CODE REVIEW

**Comment by r-barnes** [comment]
This PR makes the value of type annotations slightly more apparent. For instance, consider:
```
    def __div__(self, other: Union[int, float]) -> float:
        if not isinstance(other, (int, float)):
            raise NotImplementedError()
```
The code currently defensively checks the input type, but can only happen at runtime. The type annotation now provides a way for users to find the problem preemptively via static analysis.

**Comment by r-barnes** [comment]
Fixed merge conflicts.

**Comment by SteveDiamond** [comment]
This seems to break the build for some reason, related to imports.

**Comment by SteveDiamond** [comment]
Also we're having some flake8 issues on our Travis CI (for some reason Travis is not showing up in PR checks). The main issue is "E252 missing whitespace around parameter equals", which seems to be from the new type signatures.

**Comment by r-barnes** [comment]
@SteveDiamond : Do you use any sort of auto-formatter? e.g. Black or AutoPep8?

**Comment by SteveDiamond** [comment]
I use black in all my other projects. (I didn't know about it when I started cvxpy.) I'll discuss with the other devs, but I could do a PR to start using black, and that would resolve this issue (and only use flake8 in CI for syntax errors).

**Comment by r-barnes** [comment]
I can fix this by hand if you prefer not convert, but I thought I'd ask before going to that work.

**Comment by SteveDiamond** [comment]
> I can fix this by hand if you prefer not convert, but I thought I'd ask before going to that work.

I don't want to put you through that trouble if it can be avoided. We'll figure out if we can switch to black in a day or two.

**Comment by SteveDiamond** [comment]
Hi @r-barnes , after internal discussion we decided that we need more time to consider a change to black. For the moment I fixed the flake8 issues in the current build manually (#1327 ).

**Comment by r-barnes** [comment]
Thanks @SteveDiamond . I have flake8 and testing working locally now, so hopefully I can avoid a repeat of that, and hopefully the CI will get back online.

I've backed out the previous version of this PR in favour of one that concentrates on adding boolean typing. Circular dependencies were causing issues with the previous version. I'll come back to it later.

**Comment by r-barnes** [comment]
Added integer and floating point annotations too.

**Comment by r-barnes** [comment]
Just checking in on this.

File: cvxpy/interface/matrix_utilities.py [hunk 1]
```diff
@@ -165,7 +165,7 @@ def from_1D_to_2D(constant):
         return constant
 
 
-def convert(constant, sparse=False, convert_scalars=False):
+def convert(constant, sparse=False, convert_scalars: bool = False):
```
Comment (SteveDiamond) [inline cvxpy/interface/matrix_utilities.py:168]: Add annotation for "sparse"?

File: cvxpy/problems/problem.py [hunk 2]
```diff
@@ -822,8 +822,8 @@ def _invalidate_cache(self) -> None:
     def _solve(self,
                solver=None,
                warm_start=True,
-               verbose=False,
-               gp=False, qcp=False, requires_grad=False, enforce_dpp=False, **kwargs):
+               verbose: bool = False,
+               gp: bool = False, qcp: bool = False, requires_grad: bool = False, enforce_dpp: bool = False, **kwargs):
```
Comment (SteveDiamond) [inline cvxpy/problems/problem.py:826]: This line looks a bit long.

**Comment by SteveDiamond** [comment]
@r-barnes sorry for the slow response! I have a couple tiny comments, otherwise LGTM.

**Comment by SteveDiamond** [comment]
@r-barnes I went ahead and merged this in. Thanks again for all your contributions!

**Comment by r-barnes** [comment]
Oh, sorry, this slipped through the cracks on my end.

**Pull request merged**